feat(style): add function to replace styles while preserving the style stack

sparkles43 opened the pull request:
Hi,

during development I came up with a little problem and I wonder whether my approach to solve it might be suitable for upstream lvgl. What's your opinion on such a function?

Best regards,
sparkles

Hints for reviewers:
If most of it is OK, I would like to ask you to have a look at the use of the transition deletion in the new function. I am a bit unsure whether that is necessary or whether it is placed at the right spot.

### Description of the feature or fix

Currently, it is only possible to add styles or remove styles in lvgl. This pull request introduces a function that allows replacing styles kept in the style array of lv objects, while preserving the order of styles.

A reason to allow such replacements over the use of removing one style and adding another is the preservation of the style stack order. While removing and adding would make the replacement style the most dominant one, it can be beneficial (for abstraction purposes) to prevent this. Imagine the following:
	- We create a new widget that has two possible sets of styles
	- These styles shall be selected internally (perhaps based on an event or timing behavior)
	- However, we want a developer to further enrich our widget by adding further styles (no matter whether local or not)
	- In case the developer adds a non-local style (e.g. a huge styling change stored in flash) it becomes impossible to replace our two sets of styles WITHOUT disturbing the additional styles set by the developer


### Checkpoints
- [x] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [x] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- n/a ~~Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.~~
- [x] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- n/a ~~If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).~~
 
Be sure the following conventions are followed:
- [x] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [x] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [x] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [x] Use typed pointers instead of `void *` pointers
- [x] Do not `malloc` into a static or global variables. Instead declare the variable in `LV_ITERATE_ROOTS` list in [`lv_gc.h`](https://github.com/lvgl/lvgl/blob/master/src/misc/lv_gc.h) and mark the variable with `GC_ROOT(variable)` when it's used. See a detaild description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [x] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [x] Widget members function must start with `lv_<modul_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [x] `struct`s should be used via an API and not modified directly via their elements.
- [x] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [x] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [x] Arguments must be named in H files too.
- [x] To register and use callbacks one of the followings needs to be followed (see a detaild description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)): 
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

kisvegabor commented:
Hi sparkles,

It's really a valid use case, although it wasn't required fro me in the practice. I usually add extra styles with `LV_STATE_USER_1/2/3/4` and add/clear these states as needed. 

Anyway, I agree that it's a useful improvement and the implementation looks good too.

Could you update the docs too?

sparkles43 commented:
Hi @kisvegabor,

I have updated the documentation and added a test case for replacements. I hope it looks fine :)

Btw.: you are right. Pretty much the same effect could be created by using user states. I guess the differences of both approaches are minor.

kisvegabor commented:
Looks great! Thank you!
